File: research/online-decoder-client/python/wav_streaming_sample.py
import json
import sys

import librosa
import numpy as np
import requests
import websocket

import settings


class StreamEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        elif isinstance(obj, np.floating):
            return float(obj)
        elif isinstance(obj, np.ndarray):
            return obj.tolist()
        else:
            return super(StreamEncoder, self).default(obj)

# issue one time token
def issue_one_time_token():
    url = f'https://{settings.API_BASE}/v1/issue_token/'
    headers = {
        'accept': 'text/html',
        'Authorization': f'Bearer {settings.API_KEY}',
        'Content-Type': 'application/json',
    }
    body = {
        'product_name': settings.PRODUCT_NAME,
        'organization_id': settings.ORGANIZATION_ID,
        'user_id': settings.USER_ID
    }
    res = requests.post(url, headers=headers, data=json.dumps(body))
    if res.ok:
        return res.content.decode('utf-8')
    else:
        print(f'{res.status_code}: {res.reason}')

    return None

def main(wav_path, out_path):
    ws = None
    try:
        # 先にwavを読み込んでおく
        wav, _ = librosa.load(wav_path, settings.MODEL_SAMPLING_RATE)
        #print(wav)
        wav = wav*32678
        wav = wav.astype(np.int16)
        #print(wav)

        x = 2730
        to_chunk= lambda wav, x: [wav[i:i+x] for i in range(0, len(wav), x)]
        
        # websocket: open
        access_token = issue_one_time_token()
        if not access_token:
            return
        ws = websocket.create_connection(f'wss://{settings.API_BASE}/ws/')
        msg =  {
            "access_token": access_token,
            "type": 'start',
            "sampling_rate": settings.MODEL_SAMPLING_RATE,
            "product_name": settings.PRODUCT_NAME,
            "organization_id": settings.ORGANIZATION_ID,
            "words": settings.WORDS,
        }
        if settings.MODEL_ALIAS:
            msg['model_alias'] = settings.MODEL_ALIAS
        if settings.MODEL_NAME:
            msg['model_name'] = settings.MODEL_NAME
        if settings.USER_ID:
            msg['user_id'] = settings.USER_ID
        ws.send(json.dumps(msg))
        
        if not ws.connected:
            print('not connected')
            return

        # streaming
        with open(out_path, 'w') as f:
            for chunk in to_chunk(wav, x):
                #print('chunk=', chunk)
                msg = {
                    "type":"streamAudio",
                    "stream":list(chunk)
                }
                #print(json.dumps(msg, cls = StreamEncoder))
                ws.send(json.dumps(msg, cls = StreamEncoder))
                result = ws.recv()
                result = json.loads(result)
                if result['type'] == 'end':
                    f.write(result['result']+ '\n')
                    print(result['result'])

            #final
            ws.send(json.dumps({"type":"final"}))

            while True:
                result = ws.recv()
                if not result:
                    continue
                result = json.loads(result)
                if result['type'] == 'final-end':
                    f.write(result['result']+ '\n')
                    print(result['result'])
                    break
                elif result['type'] == 'final-decoding':
                    break
    
    except Exception as e:
        print(e)
    finally:
        if ws and ws.connected:
            ws.close()

if __name__ == '__main__':
    wav_path = sys.argv[1]
    out_path = sys.argv[2]
    main(wav_path, out_path)

File: research/online-decoder-client/nodejs/wav_streaming_sample.js
'use strict'

var fs = require('fs');

const {
  API_BASE,
  MODEL_NAME,
  MODEL_ALIAS,
  SAMPLE_RATE_HERTZ,
  PRODUCT_NAME,
  ORGANIZATION_ID,
  USER_ID,
  API_KEY,
  OUTPUT_FILE,
  WORDS
} = require('./const')

// issue_one_time_token
async function issueToken() {
  const options = {
    method: 'POST',
    uri: `https://${API_BASE}/v1/issue_token/`,
    headers: {
      'accept': 'text/html',
      'Authorization': `Bearer ${API_KEY}`,
      'Content-type': 'application/json'
    },
    body: {
      product_name: PRODUCT_NAME,
      organization_id: ORGANIZATION_ID,
      user_id: USER_ID
    },
    json: true
  }
  let token = null
  const rp = require('request-promise')
  await rp(options)
    .then(res => {
      token = res
    })
    .catch(err => {
      console.error(err)
    })
  return token
}

async function main() {
  if(process.argv.length != 4) {
    console.error("Invalid number of arguments. Required: wav_file max_bytes_per_chunk")
    process.exit(1)
  } 
  const wav_file = process.argv[2]
  const highWaterMark = parseInt(process.argv[3])
  if(highWaterMark < 1024) {
    console.error("max_bytes_per_chunk is invalid. It must be 1024 or higher.")
    process.exit(1)
  }

  const accessToken = await issueToken()
  if (accessToken === null) {
    console.error('Token could not be issued.')
    return
  }

  setInterval(() => {
  }, 5000)

  // websocket: opne
  try {
    const WebSocket = require('ws')
    const ws = new WebSocket(`wss://${API_BASE}/ws/`)
    ws.onopen = function() { 
      let msg = {
        access_token: accessToken,
        type: 'start',
        sampling_rate: SAMPLE_RATE_HERTZ,
        product_name: PRODUCT_NAME,
        organization_id: ORGANIZATION_ID,
        words: WORDS
      }
      if (MODEL_ALIAS) {
        msg.model_alias = MODEL_ALIAS
      }
      if (MODEL_NAME) {
        msg.model_name = MODEL_NAME
      }
      if (USER_ID) {
        msg.user_id = USER_ID
      }
      var m = JSON.stringify(msg);
      //console.log(m);
      ws.send(m);
    }
    
    fs.writeFileSync(OUTPUT_FILE, '', (err) => {
      if (err) throw err
    })
    ws.onmessage = function (event) {
      //console.log(`ws.onmessage ${JSON.stringify(event.data)}`)
      const res = JSON.parse(event.data)
      //console.log(`ws.onmessage ${res.result}`)
      if (res.type === 'end') {
        fs.appendFile(OUTPUT_FILE, `${res.result}\n`, (err) => {
          if (err) console.error(err)
        })
        console.log(res.result)
      }
    }

    var wav = require('wav');
     
    var file = fs.createReadStream(wav_file,  {'highWaterMark': highWaterMark});
    var reader = new wav.Reader();
     
    // the "format" event gets emitted at the end of the WAVE header
    reader.on('format', function (format) {
       console.log(`Format: ${JSON.stringify(format)}`); 
    });
     
    // pipe the WAVE file to the Reader instance
    ws.on('open', () => {
        console.log('ws open')
        file.pipe(reader);

        reader.on('data', chunk => { 
            //console.log(`got reader data len=${chunk.length} ws.readyState=${ws.readyState}`)
            var l = chunk.length
              var buf = []
              for (var i = 0; i < l / 2; i += 1) {
                buf[i] = chunk.readInt16LE(i*2, true)
                //if (buf[i] > 1.0) buf[i] = 1.0
                //else if (buf[i] < -1.0) buf[i] = -1.0
              }
              var bufferArray =  Array.prototype.slice.call(buf)
              //console.log("length", bufferArray.length)
              var msg = {
                type: 'streamAudio',
                stream: bufferArray
              }
              //console.log(bufferArray)
              var m = JSON.stringify(msg);
              //console.log(m);
              ws.send(m)
          })
          .on('error', err => {
            console.error('reader threw an error:', err)
          })
     })
  } catch (err) {
    console.error(err)
    process.exit()
  }

}

main()
